src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +198,90 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
                 throw new ArgumentOutOfRangeException(nameof(serverAddresses));
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            Dictionary<Task, Socket> socketTasks = new Dictionary<Task, Socket>(serverAddresses.Length);
+
+            ExceptionDispatchInfo lastSocketConnectEx = null;
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port);
+                    socketTasks.Add(socketTask, socket);
                 }
                 catch (Exception exc)
                 {
                     socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    lastSocketConnectEx = ExceptionDispatchInfo.Capture(exc);
+                }
+            }
+
+            if (socketTasks.Count == 0)
+            {
+                if (lastSocketConnectEx != null)
+                {
+                    lastSocketConnectEx.Throw();
+                }
+                else
+                {
+                    // Should never get here. Either there will have been no addresses and we'll have thrown
+                    // at the beginning, or one of the addresses will have worked and been added to the list, or
+                    // at least one of the addresses will have failed, raising an exception.
+                    throw new ArgumentException();
+                }
+            }
+
+            // Wait on all connection tasks, and return socket associated with first completed task.
+            Socket resultSocket = null;
+            ExceptionDispatchInfo connectWaitEx = null;
+            try
+            {
+                while (socketTasks.Count > 0)
```

Comment:
Instead of looping over all the tasks, can't we create a continuation task for socket connection which checks if the result of the ConnectAsync was successful and if it was, update the `resultSocket` and invoke the cancellation of all the other pending tasks.